From 80d73a2df2c33212217c59783f62d74673776ebb Mon Sep 17 00:00:00 2001
From: Kiall Mac Innes <kiall@managedit.ie>
Date: Fri, 4 Nov 2011 19:43:50 +0000
Subject: [PATCH 34/40] Fixes bug 834633 - Auto assigning floating IPs

Details:

 - This fix is specific to the auto_assign_floating_ip=true feature, which was new in diablo[1] and looks like it never actually worked.
 - Trey's refactoring on master[2] fixed bug #834633 where a FloatingIp object was passed to network_api.associate_floating_ip() instead of an address string.
 - In bug #834633, Serguei Koubli had previously (i.e. before Trey's refactoring) proposed a much more isolated fix for this issue.
 - This commit includes Serguei Koubli's fix, and, has also fixed a seperate issue where auto-assigned flating IPs could not be deallocated. (This fix was also included in Trey's refactoring)

[1] https://blueprints.launchpad.net/nova/+spec/floating-ip-auto-assignment
[2] https://review.openstack.org/628

Change-Id: Ie439c6743f066f8504ef88de861dee8a7762940d
---
 Authors                 |    1 +
 nova/network/manager.py |    8 +++++---
 2 files changed, 6 insertions(+), 3 deletions(-)

diff --git a/Authors b/Authors
index 2298120..cadfb26 100644
--- a/Authors
+++ b/Authors
@@ -72,6 +72,7 @@ masumoto<masumotok@nttdata.co.jp>
 Ken Pepple <ken.pepple@gmail.com>
 Kevin Bringard <kbringard@attinteractive.com>
 Kevin L. Mitchell <kevin.mitchell@rackspace.com>
+Kiall Mac Innes <kiall@managedit.ie>
 Kirill Shileev <kshileev@gmail.com>
 Koji Iida <iida.koji@lab.ntt.co.jp>
 Loganathan Parthipan <parthipan@hp.com>
diff --git a/nova/network/manager.py b/nova/network/manager.py
index c582d4d..d8de396 100644
--- a/nova/network/manager.py
+++ b/nova/network/manager.py
@@ -234,8 +234,8 @@ class FloatingIP(object):
 
             # call to correct network host to associate the floating ip
             self.network_api.associate_floating_ip(context,
-                                              floating_ip,
-                                              fixed_ip,
+                                              floating_ip['address'],
+                                              fixed_ip['address'],
                                               affect_auto_assigned=True)
         return ips
 
@@ -257,7 +257,9 @@ class FloatingIP(object):
             # disassociate floating ips related to fixed_ip
             for floating_ip in fixed_ip.floating_ips:
                 address = floating_ip['address']
-                self.network_api.disassociate_floating_ip(context, address)
+                self.network_api.disassociate_floating_ip(context,
+                                                          address,
+                                                          True)
                 # deallocate if auto_assigned
                 if floating_ip['auto_assigned']:
                     self.network_api.release_floating_ip(context,
-- 
1.7.6.4

